Initial implementation

gwicke opened the pull request:
Bug: [T173447](https://phabricator.wikimedia.org/T173447)

gwicke commented:
Initial implementation of the basic logic & interface.

TODO:
- Actually implement the redis get / put interface & test it.

Pchelolo commented on README.md:
```diff
@@ -0,0 +1,61 @@
+# budgeteer
+Cost-based request rate limiter, scheduler, and deduplicator. Uses a Redis
```
s/request/event ? It's pretty generic so it's not really focused on requests, we can limit any kind of activity can't we?

gwicke replied:
Done in https://github.com/wikimedia/budgeteer/pull/1/commits/1dfefa8e5d482f6b04350c7d9d79530ca5ba3b19.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,138 @@
+const P = require('bluebird');
+const redis = require('redis');
+
+class RedisClient {
```
I don't think we should depend on Redis here. I think it's better to settle on a get/put interface and accept an object that implements the interface - it that way we can swap reds for anything else (cassandra/memcached/zookeeper/etc)

It will be like the `mediawiki-title` that have no dependencies.

We can still provide the storage backends as a part of the module if we need, but make the client either use them explicitly or implement their own one.

gwicke replied:
Agreed. I already moved in that direction in the redis patch. The member is now called _store, and I added high-level put/get/close members in the Redis wrapper. Just need to add some config-driven backend selection logic in the constructor & move out the redis store, and we should be generic.

gwicke replied:
Moved out RedisStore in https://github.com/wikimedia/budgeteer/pull/1/commits/0694dff680a98b301e281a24868ea9b695b98979.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,138 @@
+        }
+
+        options = Object.assign(options, {
+            no_ready_check: true // Prevents sending unsupported info command to nutcracker
```
This is exactly why I'm concerned with having the `RedisClient` implementation in here - someone might not use the twemproxy and want the ready checks.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,138 @@
+        options = Object.assign(options, {
+            no_ready_check: true // Prevents sending unsupported info command to nutcracker
+        });
+        // TODO: Wait for startup?
```
Can't really wait for startup because of twemproxy

gwicke replied:
I was thinking mainly about some kind of test on whether redis works. Perhaps we could issue a read for a random key, and check whether a regular "not found" response is returned?

Pchelolo replied:
There's a bunch of events on the client https://www.npmjs.com/package/redis which we could use IFF they work with nutcracker.

gwicke replied:
Okay, lets make a note of this for a follow-up.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,138 @@
+            // If we can't connect to redis - don't worry and don't fail,
+            // just log it and ignore.
+            options.log('error/redis', e);
+        });
```
We're missing the `close` of the connection - one more reason to implement the client in ChangeProp - we can use hyper switch lifecycle there

gwicke replied:
I added a close method in the follow-up patch.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,138 @@
+        this._redis = new RedisClient(conf);
+    }
+
+    check(key, limitConf, jobTimestamp = Date.now(), cost = 0) {
```
So we're dropping node 4 officially? (default parameters)

Pchelolo commented on README.md:
````diff
@@ -0,0 +1,61 @@
+## Usage example
+
+```javascript
+const budgeteer = new Budgeteer({ host: 'redishost', port: 12345 });
````
So the idea is to create one per request as you were going to do before? What about redis connection overhead?

gwicke replied:
No, the interface is now one instance for everything. Even in the previous version, there was only one redis client.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,154 @@
+        }
+        return this._getInfo(key, limitConf)
+        .then((info) => {
+            if (info.last_success > jobTimestamp) {
```
Hm.. As I remember the algorithm, the `jobTimestamp` is actually the `rootJobTimestamp`.. A bit ambiguous..

gwicke replied:
rootJobTimestamp is MediaWiki-specific, while this library is generic. However, given that we are using "events" now, I should probably rename it to eventTimestamp.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,154 @@
+            info.last_success = startTime;
+            info.token_balance = this._calculateBalance(info, limitConf, cost);
+            info.is_scheduled = false;
+            return this._putInfo(key, info);
```
So you decided to go with no CAS? A little error because of some race in balances and `last_success` wouldn't be a bit deal, but I'm worried about `is_scheduled` and whether it's possible to false-deduplicate and loose something... I can't see cases like this right away, but I'll grab a piece of paper to think a bit deeper.

gwicke replied:
Yeah, I went for a quick read-before-write instead. This makes the window of overlap very short, as we are not waiting around for the job to finish executing.

The only scenario where I was worried about false deduplication was when a job is marked as scheduled, and execution was started. However, the code allows execution whenever the balance is positive, scheduled or not. This means that any job arriving after the budget turned positive will execute.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    _putInfo(key, info) {
+        // TODO: Use something more efficient than JSON.
```
Perhaps just `${last_success}/${is_scheduled ? 1 : 0}/${token_balance}`?

gwicke replied:
A JSON array would do almost as well. The bigger gain would come from encoding the numbers as binary rather than decimal. Could try msgpack.

Pchelolo commented on lib/budgeteer.js:
```diff
@@ -0,0 +1,154 @@
+                // Exponential back-off once we are in debit, but no more than
+                // a maximum delay.
+                newBalance = Math.max(newBalance * 2,
+                    -1 * (limitConf.max_delay_days || 7) * limitConf.token_budget_per_day);
```
max_delay_days is not documented although it seems required.

gwicke replied:
Yeah, good catch. It is not required (defaulted to 7), but needs documentation.

gwicke replied:
Added in the README.

gwicke commented:
@Pchelolo: I think I addressed all your points.

I went over the interface & especially the budget configuration again, considering how different use cases will be modeled:

## Regular event processing with retry

As in example. Doubtful about the need for budget options `max_delay_days`, and especially `initial_token_balance`. Just handle this with the general storage TTL & init with `token_budget_per_day`?

## Delayed execution

As the client controls all aspects of cost assignment and rates, it can calculate the cost that will trigger the right delay: `cost = target_delay_seconds / 86400 * token_budget_per_day`. To record the job as scheduled, the client will then call `budgeteer.reportScheduled(key, limitConf, cost)`, after pushing the job into a delay queue with a target delay less or equal the desired delay. When the job returns from this delay queue, the cost will still be negative, and resulting the job to be re-scheduled for the remaining time. This can repeat until the job is eligible for execution.

Pchelolo commented on README.md:
```diff
@@ -0,0 +1,69 @@
+    }
+    if (res.delay) {
+        // add event to a (Kafka) delay queue, based on the suggested delay.
+        return enqueueToDelayQueue(evnt, delay)
```
Typo! `evnt`

gwicke replied:
No, shortening to avoid the `event` keyword.

gwicke commented:
Simplified the interface as discussed in b20754a. Also shortened the rate parameter.

gwicke commented:
@d00rman @eevans @Pchelolo: Any final thoughts on the interface before we go ahead & merge this?

d00rman commented:
LGTM :+1:
